Source/WebKit/UIProcess/API/gtk/WebKitWebViewBase.cpp

```diff
@@ -417,7 +417,9 @@ static GtkAccessible* webkitWebViewBaseAccessibleGetFirstAccessibleChild(GtkAcce
 static void
 webkitWebViewBaseAccessibleInterfaceInit(GtkAccessibleInterface* iface)
 {
+#ifdef GTK_ACCESSIBILITY_ATSPI
     iface->get_first_accessible_child = webkitWebViewBaseAccessibleGetFirstAccessibleChild;
+#endif
```

Comment:
I think something like what Carlos mentioned would work, without wornings. Basically:

```c
#if USE(GTK4)

#ifdef GTK_ACCESSIBILITY_ATSPI
static void webkitWebViewBaseAccessibleInterfaceInit(GtkAccessibleInterface*);
WEBKIT_DEFINE_TYPE_WITH_CODE(WebKitWebViewBase, webkit_web_view_base, GTK_TYPE_WIDGET,
    G_IMPLEMENT_INTERFACE(GTK_TYPE_ACCESSIBLE, webkitWebViewBaseAccessibleInterfaceInit))
#else
WEBKIT_DEFINE_TYPE(WebKitWebViewBase, webkit_web_view_base, GTK_TYPE_WIDGET)
#endif

[... GTK3 defines ...]

#endif
```

Then this section would look like:

```c
#if USE(GTK4) && defined(GTK_ACCESSIBILITY_ATSPI)
static GtkAccessible* webkitWebViewBaseAccessibleGetFirstAccessibleChild(GtkAccessible* accessible)
{
    WebKitWebViewBase* webView = WEBKIT_WEB_VIEW_BASE(accessible);

    if (webView->priv->socketAccessible)
        return GTK_ACCESSIBLE(g_object_ref(webView->priv->socketAccessible.get()));

    if (auto* widget = gtk_widget_get_first_child(GTK_WIDGET(webView)))
        return GTK_ACCESSIBLE(g_object_ref(widget));

    return nullptr;
}

static void
webkitWebViewBaseAccessibleInterfaceInit(GtkAccessibleInterface* iface)
{
    iface->get_first_accessible_child = webkitWebViewBaseAccessibleGetFirstAccessibleChild;
}
#endif

Reply:
Applying this patch I get the following build error:

```
/home/buildbot/worker/GTK-Linux-64-bit-Release-Ubuntu-LTS-Build/build/Source/WebKit/UIProcess/API/gtk/WebKitWebViewBase.cpp:387:13: warning: ‘void webkitWebViewBaseAccessibleInterfaceInit(GtkAccessibleInterface*)’ declared ‘static’ but never defined [-Wunused-function]
  387 | static void webkitWebViewBaseAccessibleInterfaceInit(GtkAccessibleInterface*);
      |             ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

Then, if I try to guard the declaration of webkitWebViewBaseAccessibleInterfaceInit with `#if USE(GTK4) && defined(GTK_ACCESSIBILITY_ATSPI)`, I get this other error:

```
In file included from /home/buildbot/worker/GTK-Linux-64-bit-Release-Ubuntu-LTS-Build/build/Source/WebKit/UIProcess/API/gtk/WebKitWebViewBase.cpp:93:
/home/buildbot/worker/GTK-Linux-64-bit-Release-Ubuntu-LTS-Build/build/Source/WebKit/UIProcess/API/gtk/WebKitWebViewBase.cpp: In function ‘GType webkit_web_view_base_get_type_once()’:
/home/buildbot/worker/GTK-Linux-64-bit-Release-Ubuntu-LTS-Build/build/Source/WebKit/UIProcess/API/gtk/WebKitWebViewBase.cpp:392:61: error: ‘GTK_TYPE_CONTAINER’ was not declared in this scope; did you mean ‘GTK_TYPE_CONSTRAINT’?
  392 | WEBKIT_DEFINE_TYPE(WebKitWebViewBase, webkit_web_view_base, GTK_TYPE_CONTAINER)
	  |                                                             ^~~~~~~~~~~~~~~~~~
/home/buildbot/worker/GTK-Linux-64-bit-Release-Ubuntu-LTS-Build/build/WebKitBuild/GTK/Release/WTF/Headers/wtf/glib/WTFGType.h:112:13: note: in definition of macro ‘_WEBKIT_DEFINE_TYPE_EXTENDED_BEGIN’
  112 |             TYPE_PARENT, \
	  |             ^~~~~~~~~~~
/home/buildbot/worker/GTK-Linux-64-bit-Release-Ubuntu-LTS-Build/build/WebKitBuild/GTK/Release/WTF/Headers/wtf/glib/WTFGType.h:42:62: note: in expansion of macro ‘_WEBKIT_DEFINE_TYPE_EXTENDED’
   42 | #define WEBKIT_DEFINE_TYPE(TypeName, type_name, TYPE_PARENT) _WEBKIT_DEFINE_TYPE_EXTENDED(TypeName, type_name, TYPE_PARENT, 0, { })
	  |                                                              ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/buildbot/worker/GTK-Linux-64-bit-Release-Ubuntu-LTS-Build/build/Source/WebKit/UIProcess/API/gtk/WebKitWebViewBase.cpp:392:1: note: in expansion of macro ‘WEBKIT_DEFINE_TYPE’
  392 | WEBKIT_DEFINE_TYPE(WebKitWebViewBase, webkit_web_view_base, GTK_TYPE_CONTAINER)
	  | ^~~~~~~~~~~~~~~~~~
ninja: build stopped: subcommand failed.
```

Reply:
This diff seems to work for me:

```diff
diff --git a/Source/WebKit/UIProcess/API/gtk/WebKitWebViewBase.cpp b/Source/WebKit/UIProcess/API/gtk/WebKitWebViewBase.cpp
index bbe3b935d7..be5781d7e9 100644
--- a/Source/WebKit/UIProcess/API/gtk/WebKitWebViewBase.cpp
+++ b/Source/WebKit/UIProcess/API/gtk/WebKitWebViewBase.cpp
@@ -384,10 +384,16 @@ struct _WebKitWebViewBasePrivate {
  */
 
 #if USE(GTK4)
+
+#ifdef GTK_ACCESSIBILITY_ATSPI
 static void webkitWebViewBaseAccessibleInterfaceInit(GtkAccessibleInterface*);
 
 WEBKIT_DEFINE_TYPE_WITH_CODE(WebKitWebViewBase, webkit_web_view_base, GTK_TYPE_WIDGET,
     G_IMPLEMENT_INTERFACE(GTK_TYPE_ACCESSIBLE, webkitWebViewBaseAccessibleInterfaceInit))
+#else
+WEBKIT_DEFINE_TYPE(WebKitWebViewBase, webkit_web_view_base, GTK_TYPE_WIDGET)
+#endif // GTK_ACCESSIBILITY_ATSPI
+
 #else
 WEBKIT_DEFINE_TYPE(WebKitWebViewBase, webkit_web_view_base, GTK_TYPE_CONTAINER)
 #endif
@@ -398,15 +404,13 @@ static void webkitWebViewBaseDidExitFullScreen(WebKitWebViewBase*);
 static void webkitWebViewBaseRequestExitFullScreen(WebKitWebViewBase*);
 #endif
 
-#if USE(GTK4)
+#if USE(GTK4) && defined(GTK_ACCESSIBILITY_ATSPI)
 static GtkAccessible* webkitWebViewBaseAccessibleGetFirstAccessibleChild(GtkAccessible* accessible)
 {
     WebKitWebViewBase* webView = WEBKIT_WEB_VIEW_BASE(accessible);
 
-#ifdef GTK_ACCESSIBILITY_ATSPI
     if (webView->priv->socketAccessible)
         return GTK_ACCESSIBLE(g_object_ref(webView->priv->socketAccessible.get()));
-#endif
 
     if (auto* widget = gtk_widget_get_first_child(GTK_WIDGET(webView)))
         return GTK_ACCESSIBLE(g_object_ref(widget));
```

Reply:
@GeorgesStavracas Thank you, this works. Since you authorized the patch I think it's better if you push the patch in a new PR (mark the PR as Unreviewed) and I will land it.

Reply:
Sure: https://github.com/WebKit/WebKit/pull/24066